mypy/plugins/dataclasses.py

```diff
@@ -333,6 +333,38 @@ def collect_attributes(self) -> Optional[List[DataclassAttribute]]:
 
             node = sym.node
             assert not isinstance(node, PlaceholderNode)
+
+            if isinstance(node, TypeAlias):
+                ctx.api.fail(
+                    (
+                        'Type aliases inside dataclass definitions '
+                        'are not supported at runtime.'
+                    ),
+                    Context(line=node.line, column=node.column)
+                )
+                # Now do our best to simulate the runtime,
+                # which treates a TypeAlias definition in a dataclass class
+                # as an instance field with a default value.
+                #
+                # Replace the `TypeAlias` node with a `Var` node, so that we do the same.
+                target, fullname = node.target, node.fullname
+                proper_target = get_proper_type(target)
+                var_type: Union[TypeType, AnyType]
+                if isinstance(proper_target, Instance):
+                    var_type = TypeType(proper_target, line=node.line, column=node.column)
+                # Something else -- fallback to Any
+                else:
+                    var_type = AnyType(TypeOfAny.from_error)
+                var = Var(name=fullname, type=var_type)
+                var.info = cls.info
+                var._fullname = fullname
+                cls.info.names[fullname] = SymbolTableNode(
+                    kind=MDEF,
+                    node=var,
+                    plugin_generated=True,
+                )
+                sym.node = node = var
```

Comment:
Modifying the symbol table in a plugin is a little dangerous, since we can apply the plugin hook multiple times, and the error would generated only on the first run, I think. This might be fine here, but it would be easier to reason about what goes on if we'd just skip processing TypeAlias nodes, similar to what we do with ClassVar below. This wouldn't match runtime behavior, but it would be okay since the definition is invalid in any case. I'm not sure if this would cause other problems, however.

Reply:
The theoretical issue with this is that skipping the processing of the node means mypy will construct an incorrect `__init__` signature. Given the following (weird) dataclass definition:

```python
@dataclass
class Foo:
    bar: TypeAlias = int
```

mypy will infer the following `__init__` signature:

```python
def __init__(self) -> None: ...
```

But the actual signature generated by the runtime is

```python
def __init__(self, bar: type[int] = ...) -> None: ...
```

But as you say, maybe this doesn't really matter that much, since we've already warned the user that using `TypeAlias` inside a dataclass definition isn't properly supported.